src/hooks/http.js

```diff
@@ -71,138 +62,79 @@ export const parseHttpRequestOptions = (options = {}, url) => {
     port,
     uri,
     host,
-    body: '', // XXX Filled by the httpRequestEndWrapper or httpRequestEmitWrapper ( / Write)
+    body: '', // XXX Filled by the httpRequestEndWrapper or httpRequestEmitBeforeHookWrapper ( / Write)
     method,
     headers: lowerCaseObjectKeys(headers),
     protocol,
     sendTime,
   };
 };
 
-export const httpRequestWriteWrapper = requestData => writeFunc =>
-  function(...args) {
-    safeExecute(() => {
-      if (isEmptyString(requestData.body) && isValidHttpRequestBody(args[0])) {
-        const encoding = isEncodingType(args[1]) ? args[1] : 'utf8';
-        const body =
-          typeof args[0] === 'string'
-            ? Buffer(args[0]).toString(encoding)
-            : args[0].toString();
-        requestData.body += body;
-      }
-    }, logger.LOG_LEVELS.INFO)();
-    return writeFunc.apply(this, args);
+export const httpRequestWriteBeforeHookWrapper = requestData =>
+  function(args) {
+    if (isEmptyString(requestData.body)) {
+      const body = extractBodyFromWriteFunc(args);
+      if (body) requestData.body += body;
+    }
   };
 
-export const httpRequestEmitWrapper = requestData => emitFunc =>
-  function(...args) {
-    safeExecute(
-      () => {
-        if (
-          isEmptyString(requestData.body) &&
-          args[1] &&
-          args[1]._httpMessage &&
-          args[1]._httpMessage._hasBody
-        ) {
-          const httpMessage = args[1]._httpMessage;
-          let lines = [];
-          if (httpMessage.hasOwnProperty('outputData')) {
-            lines = httpMessage.outputData[0].data.split('\n');
-          } else if (httpMessage.hasOwnProperty('output')) {
-            lines = httpMessage.output[0].split('\n');
-          }
-          if (lines.length > 0) {
-            requestData.body += lines[lines.length - 1];
-          }
-        }
-      },
-      'Parse data from emit event failed',
-      logger.LOG_LEVELS.INFO
-    )();
-    return emitFunc.apply(this, arguments);
+export const httpRequestEmitBeforeHookWrapper = (requestData, requestRandomId) => {
+  const oneTimerEmitResponseHandler = runOneTimeWrapper(
+    createEmitResponseHandler(requestData, requestRandomId)
+  );
+  return function(args) {
+    if (args[0] === 'response') {
+      oneTimerEmitResponseHandler(args[1]);
+    }
+    if (args[0] === 'socket') {
+      if (isEmptyString(requestData.body)) {
```

Comment:
are you sure that you need to take the body from the socket only if the body is empty?
can't it have some like sending to the socket in batching?